src/gfx.cpp

```diff
@@ -1445,20 +1438,75 @@ void RedrawScreenRect(int left, int top, int right, int bottom)
 	VideoDriver::GetInstance()->MakeDirty(left, top, right - left, bottom - top);
 }
 
+static std::vector<Rect> _dirty_viewport_occlusions;
+static const Viewport *_dirty_viewport;
+static NWidgetDisplay _dirty_viewport_disp_flags;
+
+static void DrawDirtyViewport(uint occlusion, int left, int top, int right, int bottom)
```

Comment:
At some point, I will be tearing apart both `DrawDirtyBlocks` and `DrawDirtyViewport` and documenting and/or refactoring them until I have a clear understanding of what they do.